Implement nice pause with exponential backoff

nsextonTT opened the pull request:
### Ticket
N/A

### Problem description
Busy wait loops with no decay logic.

### What's changed
Implement exponential backoff (with a cap) and intrinsic sleep hints for busy wait loops to have less contention.

Currently soliciting feedback.

## Benchmarks:

### Method
* Installed a kernel module onto a local machine for getting the fine-grained CPU power draw.
* Created 3 benchmark functions locally. All check for an atomic flag to be flipped by a secondary thread. The only differences are Raw Spin, Spin with `ttsl::pause`, and Spin with `ttsl::nice_spin_until`.
* Benchmark call wrapped in a utility which gets the cpu-socket total energy consumption and timestamps.
* Test cases run for 30 seconds, 5x back-to-back. Highest test and lowest test are dropped. Mean reported of the remaining 3 tests. Idle test done with a `sleep 30` call. Values reported to 3 significant digits.

### Data and Results
Test variance was < 1W for each category, even without dropping the "outliers"

| Test          | Sleep 30 | Raw Spin | TT_PAUSE spin | TT_NICE_SPIN_UNTIL |
| -             | -        | -        |       -       | -  |
| Avg Power (W) | 22.6 | 46.4 | 41.3 | 30.5 |
| P - P_Idle (W)   | - | 23.8 | 18.8 | 7.94 |
| Host Power Savings | - | - | **10.9%** | **34.3%** |

During times we are otherwise busy-waiting, this PR can reduce host side power consumption by double digit percent. 

### Checklist
- [x] Single Card Perf https://github.com/tenstorrent/tt-metal/actions/runs/22084285818

#### old

- [x] [![All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml/badge.svg?branch=nsexton/0-mm_pause)](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:nsexton/0-mm_pause)
- [x] [![Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml/badge.svg?branch=nsexton/0-mm_pause)](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:nsexton/0-mm_pause)
- [x] [![cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml/badge.svg?branch=nsexton/0-mm_pause)](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:nsexton/0-mm_pause)
- [x] New/Existing tests provide coverage for changes


#### Model tests

- [x] [![(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml/badge.svg?branch=nsexton/0-mm_pause)](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:nsexton/0-mm_pause)
  - [x] `models-mandatory` preset (runs:
      - [x] [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/20653138170) and
      - [x] [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/runs/20701876264))
      - [x] [Frequent model and ttnn tests with tracy](https://github.com/tenstorrent/tt-metal/actions/runs/20701888588)
  - [x] `models-extended` preset (runs:
      the mandatory tests, plus 
      - [x] [Demo](https://github.com/tenstorrent/tt-metal/actions/runs/20700955757)
      - [x] [Model perf](https://github.com/tenstorrent/tt-metal/actions/runs/20701903088) tests)
  - [ ] other selection - specify runs

- [ ] [![(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml/badge.svg?branch=nsexton/0-mm_pause)](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:nsexton/0-mm_pause)
  - [x] `models-mandatory` preset (runs:
      - [x] [Unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/20700985825))
  - [ ] `models-extended` preset
      - [ ] [Demo](https://github.com/tenstorrent/tt-metal/actions/runs/20700988207) and
      - [ ] [Model perf](https://github.com/tenstorrent/tt-metal/actions/runs/20700989875) tests)
        - Falcon 7b: 25-80% perf improvement on kv cache ops
        - Resnet: 30-33% perf improvement on batch_size128_224x224_batchsize128's
         - test_perf_vision_cross_attention_transformer.py ::test_perf_gemma_vision[wormhole_b0-15-1-8-device_params0] - AssertionError: Failed because it's too fast
         - gemma3 op-to-op had a perf threshold failure by a tiny amount.
  - [ ] other selection - specify runs

- [ ] [![(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml/badge.svg?branch=nsexton/0-mm_pause)](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:nsexton/0-mm_pause)
  - [x] `models-mandatory` preset (runs:
      [Quick tests](https://github.com/tenstorrent/tt-metal/actions/runs/20701007298))
  - [ ] `models-extended` preset
      - [ ] [Demo](https://github.com/tenstorrent/tt-metal/actions/runs/20701008863) and
      - [ ] [Model perf](https://github.com/tenstorrent/tt-metal/actions/runs/20701010134) tests)
  - [ ] other selection - specify runs

nsextonTT commented:
/codeowners ping

cfjchu reviewed (commented):
neat! I agree with some of the other comments made on the PR as well

nsextonTT commented:
@cfjchu Spinning 100 every tap vs relying only on the sleeps (N_SPINS=0) had no measurable effect on power consumption.

bbradelTT commented:
> @bbradelTT I ran "Frequent model and ttnn tests" with and without tracing. With and without tracing was at parity with main, or at least within the same performance threshold bands.
> 
> T3K model perf on Falcon and Resnet were significantly better (list in the PR description).

Thanks @nsextonTT

riverwuTT commented on tt_stl/tt_stl/tt_pause.hpp:
```diff
@@ -0,0 +1,67 @@
+ * @param predicate A callable that returns true when the wait should end
+ */
+template <uint32_t N_SPINS = 100, uint32_t MAX_WAIT_US = 16, typename... Ts>
+__attribute__((flatten)) inline void TT_NICE_SPIN_UNTIL(auto predicate, Ts&&... args) {
```
I don't think this is a good argument. Plus this would make sense for `TT_PAUSE`, but not so much for `nice_spin_until`.

cfjchu reviewed (approved):
It's a nice change - thanks :)

riverwuTT reviewed (approved):
Thanks!

nsextonTT committed:
Implement nice pause with exponential backoff

Restore log_trace to event

Simplify backoff logic

Add unit test for tt_pause

Add riscv_pause and tidy spin logic

more rigerous pause instruction handling

implement feedback

nsextonTT committed:
clang-tidy

nsextonTT committed:
more clang tidy
